### PR TITLE
Feature/allow to hash/encrypt credentials or use an external storage 6404

### LINKED ISSUE #6404: Allow to hash API credentials
This sources from the broken feature in 2.8.2 and requires a new specification and possible sponsors.

Tasks to consider:

- [ ] Common hash algorithm (#6278)
- [ ] Creation via CLI command (api user, api setup)
- [ ] Default handling of hashes or plain text (don't do this in OnConfigLoaded with OpenSSL functions)
- [ ] Discuss this idea as password store with other feature requests

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #6404

### CODE REVIEW

File: lib/remote/apiuser.ti [hunk 1]
```diff
@@ -27,5 +28,4 @@ validator ApiUser {
 		};
 	};
 };
-
-}
+}
```
Comment (Al2Klimov) [inline lib/remote/apiuser.ti]: What's this change for?

File: lib/remote/apiuser.ti [hunk 2]
```diff
@@ -15,6 +15,7 @@ class ApiUser : ConfigObject
 	[deprecated, config, no_user_view] String password_hash;
 	[config] String client_cn (ClientCN);
 	[config] array(Value) permissions;
+	[no_user_view, no_user_modify] String hashed_password;
```
Comment (Al2Klimov) [inline lib/remote/apiuser.ti]: Please just after `password_hash`.

File: lib/remote/apiuser.ti [hunk 3]
```diff
@@ -13,6 +13,7 @@ class ApiUser : ConfigObject
 	/* No show config */
 	[no_user_view, no_user_modify] String password;
 	[deprecated, config, no_user_view] String password_hash;
+    [no_user_view, no_user_modify] String hashed_password;
```
Comment (Al2Klimov) [inline lib/remote/apiuser.ti]: Not indented like the others.

File: lib/remote/apiuser.cpp [hunk 4]
```diff
@@ -47,9 +47,18 @@ ApiUser::Ptr ApiUser::GetByAuthHeader(const String& auth_header)
...
+        if (hashedPassword.IsEmpty()) {
+            if (Utility::ComparePasswords(password, user->GetPassword())) {
+                return nullptr;
+            }
+        } else {
+            if (strcmp(hashedPassword.CStr(), crypt(password.CStr(), hashedPassword.CStr())) == 0) {
+                return nullptr;
+            }
+        }
+    }
```
Comment (Al2Klimov) [inline lib/remote/apiuser.cpp]: Not indented like the other lines.

Comment (Al2Klimov) [reply lib/remote/apiuser.cpp]: Ask @N-o-X about CLion + tabs/spaces.

File: lib/remote/apiuser.cpp [hunk 5]
```diff
@@ -47,9 +47,18 @@ ApiUser::Ptr ApiUser::GetByAuthHeader(const String& auth_header)
...
-		return nullptr;
-
+	else if (user) {
+        String hashedPassword = user->GetHashedPassword();
+        if (hashedPassword.IsEmpty()) {
+            if (Utility::ComparePasswords(password, user->GetPassword())) {
+                return nullptr;
+            }
+        } else {
+            if (strcmp(hashedPassword.CStr(), crypt(password.CStr(), hashedPassword.CStr())) == 0) {
```
Comment (dnsmichi) [inline lib/remote/apiuser.cpp]: Please move this into a dedicated function, e.g. Utility::CompareHashedPasswords() which returns a boolean value.
The next step is to add unit tests for this.

Comment (Al2Klimov) [reply lib/remote/apiuser.cpp]: ... once it works at all.

File: lib/remote/apiuser.cpp [hunk 6]
```diff
@@ -47,9 +47,16 @@ ApiUser::Ptr ApiUser::GetByAuthHeader(const String& auth_header)
 	 */
 	if (!user || password.IsEmpty())
 		return nullptr;
-	else if (user && !Utility::ComparePasswords(password, user->GetPassword()))
-		return nullptr;
-
+	else if (user) {
```
Comment (Al2Klimov) [inline lib/remote/apiuser.cpp]: Didn't @N-o-X tell you that this condition is a no-op?

File: lib/remote/apiuser.cpp [hunk 7]
```diff
@@ -47,9 +47,16 @@ ApiUser::Ptr ApiUser::GetByAuthHeader(const String& auth_header)
...
+		String hashedPassword = user->GetHashedPassword();
+		if (hashedPassword.IsEmpty()) {
+			if (Utility::ComparePasswords(password, user->GetPassword()))
+				return nullptr;
+		} else {
+			if (strcmp(hashedPassword.CStr(), crypt(password.CStr(), hashedPassword.CStr())) == 0)
+				return nullptr;
+		}
+	}
 	return user;
```
Comment (Al2Klimov) [inline lib/remote/apiuser.cpp]: Separate these two lines by an empty line.

File: lib/remote/apiuser.cpp [hunk 8]
```diff
@@ -47,9 +47,16 @@ ApiUser::Ptr ApiUser::GetByAuthHeader(const String& auth_header)
 	 */
 	if (!user || password.IsEmpty())
 		return nullptr;
-	else if (user && !Utility::ComparePasswords(password, user->GetPassword()))
-		return nullptr;
-
+	else if (user) {
+		String hashedPassword = user->GetHashedPassword();
+		if (hashedPassword.IsEmpty()) {
```
Comment (Al2Klimov) [inline lib/remote/apiuser.cpp]: Separate these two lines by an empty line.

**Comment by Al2Klimov** [comment]
@justinmueller05082000 Hm, Your Windows CI seems not to run for PRs. Please set up a Windows dev VM (w/o product key, yet) as documented in the Icinga 2 docs and build this branch.

**Comment by justinmueller05082000** [comment]
@Al2Klimov If I try to build the branch I get the following error. 

Screenshot from 2020-03-11 12-55-05

**Comment by justinmueller05082000** [comment]
I unsuccessfully tried to find a library for crypt to include. I also tried to find alternatives for crypt to use but I had no success with that. My next step would be to install the library on my VM, open it with VS and then try to build icinga2 again.

**Comment by Al2Klimov** [comment]
I'm afraid you won't be able to build glibc on WinNT. IMAO we should re-evaluate what hasher to use, but start with the cross-platform-availability, not `crypt(3)` compatibility.

**Pull request closed without merging**